### ISSUE TITLE
Keep short promise chains indented

### ISSUE THREAD

**Issue by n1k0** [prettier/prettier#3594]
**Prettier 1.9.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzCEA6VMwHQwAWCAFCQJQAEAvAHyUBGANgIaEXna4HFRlV1GrduXIgANCAgAHGAEtoAZ2SgWAJzUQA7gAV1CZShYA3CHIAmEkC0UxkqFk0VxJDNSzABrODADK0jzkoAHNkGDUAVxcQcwgwe0dnSSDnNRgdd2CAWxYEp2iAK0UADwAhdy8fXxYsuAAZILg8pKkImGk2gCZm6IC1VOQQBhYGAE8maCtpNSCYAHULImQADgAGSWmIZzn3aUHpuFTjJsk1OABHCLkzjJZs3KQHfMlnLLkwyOjFIOCmOABFCIQeA9SQwEYLcxLJCdMHuORMH4AYQgWRygyg0BOIAizgAKiNDE9nABfElAA)

**Input:**
```jsx
foo
  .then(() => blah())
  .then(() => blah())
```

**Output:**
```jsx
foo.then(() => blah()).then(() => blah());
```

**Expected behavior:**

I'd expect the indentation intent (for obvious legibility purpose) to be kept as is.

(Sorry if that's a duplicate)

**Comment by j-f1** [comment]
FYI, Prettier’s docs [say]() this:

> Prettier … strives to fit the most code into every line.

**Comment by azz** [comment]
@j-f1 That's true, and that's the way Prettier's core printing algorithm works, but we have also have to print readable code. That's why we did #2673.

I feel like the case posted wouldn't happen too often in regular code, as the variable are really short with no arguments. E.g. if it was a bit longer, it would expand:

**Prettier 1.9.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAzOMwAWcAJpgLw56EAUAOiAPTYQT0CUA3LVLvkcd1owAdDCJRqAJzhoADujjkAfJmlyFwgFZpo1Nm0EixCatujLMsyRDAy0AFQIBLKAHM0pnVGHEAhjF9RZzc0fS4oEAAaEAhZGCd0ZFBfSWsAdwAFFIQ0ZBBfADcIJ2Io-IxkbF8AGzQ4aIAjSV8wAGs8AGVZFpdXZBhJAFd6kGIbSpq66Jc6yRgM5tcAW18J2pHtAA8AIWa2zt8luAAZFzg1qZjBmFlrgCYLke7JWbyG3waAT2roMqsXGAAdRKYmQAA4AAzRKwQOqA5qyPJWGRwSQFc7RaQAR0GTmkC18y1WSCq62idSWTn6QxGaF61TgAEVBhB4I9ogEGsDiKCkHcOc0nNVegBhCBLFZ5KDQDEgQZ1ewfXIkyZwAC+aqAA)

**Input:**
```jsx
const fetched = fetch("/foo");
fetched
	.then(response => response.json())
	.then(json => processThings(json.data.things));

```

**Output:**
```jsx
const fetched = fetch("/foo");
fetched
  .then(response => response.json())
  .then(json => processThings(json.data.things));

```

**Comment by n1k0** [comment]
My main personal concern here is it breaks developer intents. Imho it's not because function names are short than they should be inlined and spoil legibility, even slightly.

I understand that detecting intent is a difficult topic though, and I can probably live with increasing my function names length ;)

**Comment by DataTables** [comment]
I've just hit this as well - here is the code I was using:

```
        let column = new Column(null, conn)
            .table(data.table)
            .json(data.column);
```

Prettier's output:

```
let column = new Column(null, conn).table(data.table).json(data.column);
```

If I were to add one more method to the chain, then, based on #2673 it does indeed format it on multiple lines, even if it still fits in the line width (just checked and indeed it does).

From #2673 it looks like 3 was used because that's what ESLint used - is that correct? Is there any scope to have it at 2 instead which would resolve this?

**Comment by lydell** [comment]
@DataTables It's not so easy as just changing the 3 to a 2 because we also have this issue the other way around, more or less: #3621

**Comment by DataTables** [comment]
Agreed - that's a bit of a nightmare! With changing the 3 to a 2 it would also have the effect in #3621 of transforming the toLowerCase line:

```js
const palindrome = str => {
  const s = str
    .toLowerCase()
    .replace(/[\W_]/g, "");
  return (
    s ===
    s
      .split("")
      .reverse()
      .join("")
  );
};
```

I'm afraid I don't know what the answer is here. In my case perhaps because it is a `new` constructor that would mean everything on it should be formatted as newline / indented chained?

Having an arbitrary number of three methods being chained feels a bit odd though. The problem is that in some cases breaking the chain on to multiple lines helps readability and in some cases not!

Digging around a little, the [AirBnB style guide](https://github.com/airbnb/javascript#whitespace--chains) says:

> 19.6 Use indentation when making long method chains (more than 2 method chains).

Interestingly, the ESLint `newline-per-chained-call` option is two or more, but they have a `ignoreChainWithDepth` option which is default 2, which perhaps is where the "3" for the AirBnB style guide comes from, or perhaps ESLint adopted it from there.

Personally I find chained functions much more readable on multiple lines.

**Comment by DataTables** [comment]
Sorry - I don't want to spam, but how about it would be new line per method if:

* `new` is used to invoke a constructor, or
* A function is used as an argument in any of the chained methods?

That would allow for the above examples of the promise chain from @n1k0, my own `new` chain and also the two string chaining from @atomiks.

**Comment by lydell** [comment]
Good ideas!

**Comment by ibrahim-delice** [comment]
Almost 5 years passed and this ticket is still open.

Any final solution on this?
